pkg/util/tracing/grpcinterceptor/grpc_interceptor.go

```diff
@@ -338,28 +339,36 @@ func newTracingClientStream(
 		if !atomic.CompareAndSwapInt32(isFinished, 0, 1) {
 			return
 		}
-		close(finishChan)
 		defer clientSpan.Finish()
 		if err != nil {
 			clientSpan.Recordf("error: %s", err)
 			setGRPCErrorTag(clientSpan, err)
 		}
 	}
-	go func() {
-		select {
-		case <-finishChan:
-			// The client span is being finished by another code path; hence, no
-			// action is necessary.
-		case <-ctx.Done():
+
+	if ctx.Done() != nil {
+		// All contexts that complete (ctx.Done() != nil) used in cockroach should support
+		// direct cancellation detection, since they should be derived from one of the
+		// standard context.Context.
+		// But, be safe and loudly fail tests in case somebody introduces strange
+		// context implementation.
+		if buildutil.CrdbTestBuild && !ctxutil.CanDirectlyDetectCancellation(ctx) {
```

Comment:
Let's move this into `WhenDone()`, I think we always want to assert this. If we ever make a custom context implementation that doesn't derive from the parent's cancellation, we can come up with something else then. Better safe than sorry.